src/lib/rpc/bestRpcTracker.ts

```diff
@@ -0,0 +1,346 @@
+    const isTrackingEnabled = (warmUp || !isUnusedChain) && providers.length > 1;
+
+    if (!isTrackingEnabled) {
+      return;
+    }
+
+    const probePromises = providers.map((providerInfo) => {
+      return probeRpc(chainId, providerInfo.provider, providerInfo.url);
+    });
+
+    const probeResults = await Promise.all(probePromises);
+    const successProbeResults = probeResults.filter((probe) => probe.isSuccess);
+
+    if (!successProbeResults.length) {
+      setCurrentProvider(chainId, getFallbackRpcUrl(chainId));
+
+      return;
+    }
+
+    const probeResultsByBlockNumber = orderBy(successProbeResults, ["blockNumber"], ["desc"]);
+    let bestBlockNumberProbe = probeResultsByBlockNumber[0];
+
+    const probeStats = probeResultsByBlockNumber.map((probe, i, arr) => {
+      let isValid = probe.isSuccess;
+
+      const bestBlockNumber = bestBlockNumberProbe.blockNumber;
+      const currProbeBlockNumber = probe.blockNumber;
+      const nextProbeBlockNumber = arr[i + 1]?.blockNumber;
+
+      // Rare case when RPC returned a block number from the far future
+      if (
+        i === 0 &&
+        currProbeBlockNumber &&
+        nextProbeBlockNumber &&
+        currProbeBlockNumber - nextProbeBlockNumber > BLOCK_FROM_FUTURE_THRESHOLD
+      ) {
+        bestBlockNumberProbe = arr[i + 1];
+        isValid = false;
+      }
+
+      // If the block number is lagging behind the best one
+      if (bestBlockNumber && currProbeBlockNumber && bestBlockNumber - currProbeBlockNumber > BLOCK_LAGGING_THRESHOLD) {
+        isValid = false;
+      }
+
+      return {
+        ...probe,
+        isValid,
+      };
+    });
+
+    const bestResponseTimeValidProbe = minBy(
+      probeStats.filter((probe) => probe.isValid),
+      "responseTime"
+    );
+
+    if (bestResponseTimeValidProbe) {
+      setCurrentProvider(chainId, bestResponseTimeValidProbe.url);
+    }
+
+    if (isDebugMode()) {
+      // eslint-disable-next-line no-console
+      console.table(
+        orderBy(
+          probeStats.map((probe) => ({
+            url: probe.url,
+            isSelected: probe.url === bestResponseTimeValidProbe?.url ? "✅" : "",
+            isValid: probe.isValid ? "✅" : "❌",
+            responseTime: probe.responseTime,
+            blockNumber: probe.blockNumber,
+          })),
+          ["responseTime"],
+          ["asc"]
+        )
+      );
+    }
+  });
+}
+
+function setCurrentProvider(chainId: number, newProviderUrl: string) {
+  if (!trackerState) {
+    throw new Error("RPC tracker state is not initialized");
+  }
+
+  trackerState[chainId].currentBestProviderUrl = newProviderUrl;
+
+  const storageKey = JSON.stringify(getRpcProviderKey(chainId));
+
+  localStorage.setItem(
+    storageKey,
+    JSON.stringify({
+      rpcUrl: newProviderUrl,
+      timestamp: Date.now(),
+    })
+  );
+}
+
+function isDebugMode() {
+  return localStorage.getItem(JSON.stringify(SHOW_DEBUG_VALUES_KEY)) === "true";
+}
```

Comment:
maybe we can move isDebugMode to some utils file?

Reply:
Moved to the `lib/localStorage`